os/kernel/preference/preference_read.c

```diff
@@ -44,59 +46,62 @@ static int preference_read_fs_key(char *path, preference_data_t *data)
 		return PREFERENCE_IO_ERROR;
 	}
 
-	/* Read and Verify the type and len of data value */
-	read_size = 2 * sizeof(int);
-	ret = read(fd, (void *)&value_data, read_size);
-	if (ret != read_size) {
+	check_crc = 0;
+
+	/* Read and Verify attributes of data : type, len */
+	ret = read(fd, (FAR uint8_t *)&attr, sizeof(value_attr_t));
```

Comment:
Below is simple logic & struct to handle read. You may understand write case after read this.
```
struct preference_entry_s {
    char *key;
    uint8_t type;
    uint32_t datalen; //is it enough size??
    uint32_t crc32;
}

# case read
int fd;
int ret;
struct preference_entry_s *entry;

fd = open(path, O_RDONLY); //0666 is not necessary, actually.
if (fd < 0) {
    dbg("error, errno : %d\n", errno);
    return error;
}
/* Verify Entry first */
int size = sizeof(struct preference_entry_s);
ret = read(fd, entry, size); 

if (ret < size) {
   dbg("read error, ret : %d errno : %d\n", ret, errno);
    return error;
}

if (type_from_user != entry->type) {
    dbg("error....blah blah\n");
    return error;
}
... //check entry here.

/* Now read data */
char *buf = malloc(entry->datalen);
ret = read(fd, buf, entry->datalen);
if (ret != entry->datalen) {
   blahblah
}

/* Check crc by compare crc value in entry and calculated read data */
if (entry->crc32 != crc32part()) {
   dbg("error");
   return error;
}
```

Anyway, now you calling read 3 times here, Please make sure that  I/O  always takes more much of time than any other operation.